### DISCUSSION TITLE
Halmet example code Analog Inputs setup

### DISCUSSION THREAD

**Discussion by Steve7848** [hatlabs/discussions#107] [General]
I am trying to setup the Halmet and have a question about the fuel level gauge. 

Do I have to put a jumper on A1 to use a resistor style fuel gauge?
I am in the United States and we typically use the 240-30 ohm fuel gauges. I have a couple questions about setup. In the example code, halmet_analog.cpp line 9, do I have to change kAnalogInputScale at all? or is this fixed for all analog inputs?  I am starting with the fuel level gauge but would like to add oil pressure and coolant temperature next.

If I set my tank size on line 15, kTankDefaultSize in cubic meters would the display on the Halmet display in percentage of full ie. 100% 50% or would it be displaying units of fuel left in the tank?

With the sensors that we typically use being 240 ohm empty and 30 ohms full, For line 54 tank level curve interpolation would my code look like what I have listed below?  (240 ohms, 0 m^3) and (30 ohms, .0568 m^3) the tank size is 15 gallons. Or would I not use m^3 and use percentage of full?

    tank_level->clear_samples();
    tank_level->add_sample(CurveInterpolator::Sample(240, 0));
    tank_level->add_sample(CurveInterpolator::Sample(30, .0568));

As a general question, is there a forum or anywhere that goes through someone setting this up start to finish on their system that I can use as a reference?

Thanks

**Comment by mairas** [comment]
If you only have HALMET connected to the sender, then you should place the jumper. If you want to connect HALMET in parallel to an existing gauge, then the gauge will provide the measurement voltage/current, and you should leave the jumper out.

Don't touch `kAnalogInputScale`. That value depends on the voltage divider that scales the analog inputs to levels acceptable by the ADS1115 analog to digital converter. Just like you figured out, you can fix the scaling using the `CurveInterpolator`. You can set both the curve interpolator samples and the tank size (Linear) scaling value through the web UI as well, so changing the code is not necessary (but of course fine, if you prefer to do it that way).

The Curve Interpolator should output the fill level ratio between 0..1. If you want to do it in code, something like this:

    tank_level->clear_samples();
    tank_level->add_sample(CurveInterpolator::Sample(10000., 0));
    tank_level->add_sample(CurveInterpolator::Sample(240., 0));
    tank_level->add_sample(CurveInterpolator::Sample(30., 1));
    tank_level->add_sample(CurveInterpolator::Sample(0., 1));

I've added two extra points to clip the output number. If the sender resistance is above 240 ohms, you'd otherwise get a negative or an unspecified value. And likewise for values less than 30 ohms. I don't remember what the exact implementation of `CurveInterpolator` is, but nevertheless, it's better to tell it which ranges map to 0 and 1, respectively.

I don't think there's a proper walkthrough of a HALMET setup anywhere yet. There are previous questions on this forum that might help here and there. I'm acutely aware that I should finish writing the documentation...
